File: trunk/ServerPhp/src/dc/DC0035.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0035 Controller: Contracts
 */

require_once(PATH_N21LIB."/Http/Controller.php");

class DC0035 extends Controller {

  public function __construct($request, $response, $session) {
    parent::__construct($request, $response, $session);
    $this->_initFields();
  }

private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_BPCONTR_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.BPCONTR_ID like :BPCONTR_ID";
      $params["BPCONTR_ID"] = $_REQUEST["QRY_BPCONTR_ID"];
    }
    if (!empty($_REQUEST["QRY_CLIENT_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.CLIENT_ID like :CLIENT_ID";
      $params["CLIENT_ID"] = $_REQUEST["QRY_CLIENT_ID"];
    }
    if (!empty($_REQUEST["QRY_CONTRSTAT_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.CONTRSTAT_CODE like :CONTRSTAT_CODE";
      $params["CONTRSTAT_CODE"] = $_REQUEST["QRY_CONTRSTAT_CODE"];
    }
    if (!empty($_REQUEST["QRY_CONTRTYPE_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.CONTRTYPE_CODE like :CONTRTYPE_CODE";
      $params["CONTRTYPE_CODE"] = $_REQUEST["QRY_CONTRTYPE_CODE"];
    }
    if (!empty($_REQUEST["QRY_CUST_BPARTNER_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.CUST_BPARTNER_ID like :CUST_BPARTNER_ID";
      $params["CUST_BPARTNER_ID"] = $_REQUEST["QRY_CUST_BPARTNER_ID"];
    }
    if (!empty($_REQUEST["QRY_DOC_DATE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.DOC_DATE like :DOC_DATE";
      $params["DOC_DATE"] = $_REQUEST["QRY_DOC_DATE"];
    }
    if (!empty($_REQUEST["QRY_DOC_NO"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.DOC_NO like :DOC_NO";
      $params["DOC_NO"] = $_REQUEST["QRY_DOC_NO"];
    }
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_SUPP_BPARTNER_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.SUPP_BPARTNER_ID like :SUPP_BPARTNER_ID";
      $params["SUPP_BPARTNER_ID"] = $_REQUEST["QRY_SUPP_BPARTNER_ID"];
    }
}

public function doQuery() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                c.BPCONTR_ID
                ,(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME
                ,c.CLIENT_ID
                ,(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME
                ,c.CONTRSTAT_CODE
                ,c.CONTRTYPE_CODE
                ,c.CREATEDBY
                ,c.CREATEDON
                ,c.CUST_BPARTNER_ID
                ,(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME
                ,c.DOC_DATE
                ,c.DOC_NO
                ,c.ENDDATE
                ,c.ID
                ,c.MODIFIEDBY
                ,c.MODIFIEDON
                ,c.NOTES
                ,c.STARTDATE
                ,c.SUPP_BPARTNER_ID
                ,(select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME
            from BP_CONTRACT c $where $orderByClause ";
    $this->logger->debug($sql);
    $rs = $this->getDbConn()->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $dataOut = $this->serializeCursor($rs,array_keys($this->fields), $this->getDataFormatFetch());
    if ($this->getDataFormatFetch() == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
} /* end function doQuery  */


public function doExport() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                c.ID
                ,c.CLIENT_ID
                ,(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME
                ,c.DOC_NO
                ,c.DOC_DATE
                ,(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME
                ,c.CUST_BPARTNER_ID
                ,c.SUPP_BPARTNER_ID
                ,(select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME
                ,c.CONTRTYPE_CODE
                ,c.STARTDATE
                ,c.ENDDATE
                ,(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME
                ,c.BPCONTR_ID
                ,c.NOTES
                ,c.CREATEDON
                ,c.CREATEDBY
                ,c.MODIFIEDON
                ,c.MODIFIEDBY
                ,c.CONTRSTAT_CODE
            from BP_CONTRACT c $where $orderByClause ";
    $rs = $this->getDbConn()->Execute($sql, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    } else {
      $columns = array_keys($this->fields);
    }
    if ($this->getDataFormatExport() == "csv" ) {
      $dataOut = $this->serializeCursor($rs,$columns,"csv");
    } else {
      $dataOut = $this->serializeCursor($rs,$columns,"xml");
      $dataOut = "<records>".$dataOut."</records>";
      $dataOut = "<queryParams>".$this->serializeMap($params,"xml")."</queryParams>".$dataOut;
      $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fields,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fields,false)."</columnDef>".$dataOut;
      $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
      $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
      $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    }
    $this->beginExport();
    print $dataOut;
    $this->endExport();
} /* end function doExport  */

public function fetchRecord() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0035.fetchRecord().");}
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function fetchRecord */

public function doInsert() {
    $this->populateRecordFromRequest(); 
    $sql = "insert into BP_CONTRACT(
                 BPCONTR_ID
                ,CLIENT_ID
                ,CONTRSTAT_CODE
                ,CONTRTYPE_CODE
                ,CREATEDBY
                ,CUST_BPARTNER_ID
                ,DOC_DATE
                ,DOC_NO
                ,ENDDATE
                ,ID
                ,MODIFIEDBY
                ,NOTES
                ,STARTDATE
                ,SUPP_BPARTNER_ID
            ) values ( 
                 :BPCONTR_ID
                ,:CLIENT_ID
                ,:CONTRSTAT_CODE
                ,:CONTRTYPE_CODE
                ,:CREATEDBY
                ,:CUST_BPARTNER_ID
                ,:DOC_DATE
                ,:DOC_NO
                ,:ENDDATE
                ,:ID
                ,:MODIFIEDBY
                ,:NOTES
                ,:STARTDATE
                ,:SUPP_BPARTNER_ID
    )";
    $stmt = $this->getDbConn()->prepare($sql);
    $_seq = $this->getDbConn()->execute("select SEQ_BPCONTR_ID.nextval seq_val from dual")->fetchRow();
    $this->record["ID"] = $_seq["SEQ_VAL"];
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doInsert */

public function doUpdate() {
    $this->populateRecordFromRequest();
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0035.doUpdate().");}
    $sql = "update BP_CONTRACT set 
                 ID=:ID
                ,DOC_NO=:DOC_NO
                ,DOC_DATE=:DOC_DATE
                ,CLIENT_ID=:CLIENT_ID
                ,CUST_BPARTNER_ID=:CUST_BPARTNER_ID
                ,CONTRTYPE_CODE=:CONTRTYPE_CODE
                ,STARTDATE=:STARTDATE
                ,ENDDATE=:ENDDATE
                ,NOTES=:NOTES
                ,BPCONTR_ID=:BPCONTR_ID
                ,CONTRSTAT_CODE=:CONTRSTAT_CODE
                ,SUPP_BPARTNER_ID=:SUPP_BPARTNER_ID
    where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"],"ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doUpdate */

public function doDelete() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0035.doDelete().");}
    $sql = "delete from BP_CONTRACT where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doDelete */

public function initNewRecord() {
    $this->populateRecordFromRequest();
    $this->record["_p_record_status"] = "insert";
    $this->setFieldInitialValues($this->record, "DC0035");
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                c.BPCONTR_ID
                ,(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME
                ,c.CLIENT_ID
                ,(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME
                ,c.CONTRSTAT_CODE
                ,c.CONTRTYPE_CODE
                ,c.CREATEDBY
                ,c.CREATEDON
                ,c.CUST_BPARTNER_ID
                ,(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME
                ,c.DOC_DATE
                ,c.DOC_NO
                ,c.ENDDATE
                ,c.ID
                ,c.MODIFIEDBY
                ,c.MODIFIEDON
                ,c.NOTES
                ,c.STARTDATE
                ,c.SUPP_BPARTNER_ID
                ,(select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME
            from BP_CONTRACT c
         where 
           c.ID= :ID
            ";
    $rs = $this->getDbConn()->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    CollectionUtils::copyMapValues($record, $row);
} /* end function findByPk  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
    $this->populateRecordFromRequest();
    print "{success:true, data:".json_encode($this->record)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
} /* end function callProcedure */


private function _initFields() {
  $this->fields = array(
  "BPCONTR_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"BPCONTR_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CLIENT_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"CLIENT_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CONTRSTAT_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CONTRTYPE_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CREATEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CREATEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"CUST_BPARTNER_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"CUST_BPARTNER_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"DOC_DATE" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"DOC_NO" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"ENDDATE" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"MODIFIEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"MODIFIEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"NOTES" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"STARTDATE" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"SUPP_BPARTNER_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"SUPP_BPARTNER_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
);
}

}
?>

File: trunk/ServerJava/src/net/nan21/ebs/dc/DC0035.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0035 DC Controller: Contracts
 */

package net.nan21.ebs.dc;


import java.util.*;
import javax.servlet.http.HttpServletResponse;
import net.nan21.lib.*;

public class DC0035 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_DOC_NO") != null && !this.request.getParam("QRY_DOC_NO").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.DOC_NO like :DOC_NO");
      this.queryParams.put("DOC_NO",(String)this.request.getParam("QRY_DOC_NO"));
    }
    if (this.request.getParam("QRY_DOC_DATE") != null && !this.request.getParam("QRY_DOC_DATE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.DOC_DATE like :DOC_DATE");
      this.queryParams.put("DOC_DATE",(String)this.request.getParam("QRY_DOC_DATE"));
    }
    if (this.request.getParam("QRY_CLIENT_ID") != null && !this.request.getParam("QRY_CLIENT_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.CLIENT_ID like :CLIENT_ID");
      this.queryParams.put("CLIENT_ID",(String)this.request.getParam("QRY_CLIENT_ID"));
    }
    if (this.request.getParam("QRY_CUST_BPARTNER_ID") != null && !this.request.getParam("QRY_CUST_BPARTNER_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.CUST_BPARTNER_ID like :CUST_BPARTNER_ID");
      this.queryParams.put("CUST_BPARTNER_ID",(String)this.request.getParam("QRY_CUST_BPARTNER_ID"));
    }
    if (this.request.getParam("QRY_CONTRTYPE_CODE") != null && !this.request.getParam("QRY_CONTRTYPE_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.CONTRTYPE_CODE like :CONTRTYPE_CODE");
      this.queryParams.put("CONTRTYPE_CODE",(String)this.request.getParam("QRY_CONTRTYPE_CODE"));
    }
    if (this.request.getParam("QRY_BPCONTR_ID") != null && !this.request.getParam("QRY_BPCONTR_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.BPCONTR_ID like :BPCONTR_ID");
      this.queryParams.put("BPCONTR_ID",(String)this.request.getParam("QRY_BPCONTR_ID"));
    }
    if (this.request.getParam("QRY_CONTRSTAT_CODE") != null && !this.request.getParam("QRY_CONTRSTAT_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.CONTRSTAT_CODE like :CONTRSTAT_CODE");
      this.queryParams.put("CONTRSTAT_CODE",(String)this.request.getParam("QRY_CONTRSTAT_CODE"));
    }
    if (this.request.getParam("QRY_SUPP_BPARTNER_ID") != null && !this.request.getParam("QRY_SUPP_BPARTNER_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.SUPP_BPARTNER_ID like :SUPP_BPARTNER_ID");
      this.queryParams.put("SUPP_BPARTNER_ID",(String)this.request.getParam("QRY_SUPP_BPARTNER_ID"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " (select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME"+
               " ,(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME"+
               " ,c.ID"+
               " ,c.DOC_NO"+
               " ,c.DOC_DATE"+
               " ,c.CLIENT_ID"+
               " ,c.CUST_BPARTNER_ID"+
               " ,c.CONTRTYPE_CODE"+
               " ,c.STARTDATE"+
               " ,c.ENDDATE"+
               " ,c.NOTES"+
               " ,c.BPCONTR_ID"+
               " ,c.CREATEDON"+
               " ,c.CREATEDBY"+
               " ,c.MODIFIEDON"+
               " ,c.MODIFIEDBY"+
               " ,c.CONTRSTAT_CODE"+
               " ,c.SUPP_BPARTNER_ID"+
               " ,(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME"+
               " ,(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME"+
           " from BP_CONTRACT c "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " c.ID"+
               ",(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME"+
               " ,c.CLIENT_ID"+
               " ,c.DOC_NO"+
               " ,c.DOC_DATE"+
               " ,c.CUST_BPARTNER_ID"+
               ",(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME"+
               " ,c.SUPP_BPARTNER_ID"+
               ",(select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME"+
               " ,c.CONTRTYPE_CODE"+
               " ,c.STARTDATE"+
               " ,c.ENDDATE"+
               ",(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME"+
               " ,c.BPCONTR_ID"+
               " ,c.NOTES"+
               " ,c.CREATEDON"+
               " ,c.CREATEDBY"+
               " ,c.MODIFIEDON"+
               " ,c.MODIFIEDBY"+
               " ,c.CONTRSTAT_CODE"+
           " from BP_CONTRACT c "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into BP_CONTRACT("+
               "  ID"+
               " ,DOC_NO"+
               " ,DOC_DATE"+
               " ,CLIENT_ID"+
               " ,CUST_BPARTNER_ID"+
               " ,CONTRTYPE_CODE"+
               " ,STARTDATE"+
               " ,ENDDATE"+
               " ,NOTES"+
               " ,BPCONTR_ID"+
               " ,CREATEDBY"+
               " ,MODIFIEDBY"+
               " ,CONTRSTAT_CODE"+
               " ,SUPP_BPARTNER_ID"+
           " ) values ( "+
               "  :ID"+
               " ,:DOC_NO"+
               " ,:DOC_DATE"+
               " ,:CLIENT_ID"+
               " ,:CUST_BPARTNER_ID"+
               " ,:CONTRTYPE_CODE"+
               " ,:STARTDATE"+
               " ,:ENDDATE"+
               " ,:NOTES"+
               " ,:BPCONTR_ID"+
               " ,:CREATEDBY"+
               " ,:MODIFIEDBY"+
               " ,:CONTRSTAT_CODE"+
               " ,:SUPP_BPARTNER_ID"+
    ")";
    this.record.put("ID",   dbm.getSequenceNextValue("SEQ_BPCONTR_ID")  );
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update BP_CONTRACT set "+
               "  ID=:ID"+
               " ,DOC_NO=:DOC_NO"+
               " ,DOC_DATE=:DOC_DATE"+
               " ,CLIENT_ID=:CLIENT_ID"+
               " ,CUST_BPARTNER_ID=:CUST_BPARTNER_ID"+
               " ,CONTRTYPE_CODE=:CONTRTYPE_CODE"+
               " ,STARTDATE=:STARTDATE"+
               " ,ENDDATE=:ENDDATE"+
               " ,NOTES=:NOTES"+
               " ,BPCONTR_ID=:BPCONTR_ID"+
               " ,CONTRSTAT_CODE=:CONTRSTAT_CODE"+
               " ,SUPP_BPARTNER_ID=:SUPP_BPARTNER_ID"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from BP_CONTRACT where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
                "(select t.name from bpartner t where t.id = c.supp_bpartner_id) SUPP_BPARTNER_NAME"+
                ",(select tt.doc_no||'/'||tt.doc_date from bp_contract tt where tt.id = c.bpcontr_id ) BPCONTR_NAME"+
               " ,c.ID"+
               " ,c.DOC_NO"+
               " ,c.DOC_DATE"+
               " ,c.CLIENT_ID"+
               " ,c.CUST_BPARTNER_ID"+
               " ,c.CONTRTYPE_CODE"+
               " ,c.STARTDATE"+
               " ,c.ENDDATE"+
               " ,c.NOTES"+
               " ,c.BPCONTR_ID"+
               " ,c.CREATEDON"+
               " ,c.CREATEDBY"+
               " ,c.MODIFIEDON"+
               " ,c.MODIFIEDBY"+
               " ,c.CONTRSTAT_CODE"+
               " ,c.SUPP_BPARTNER_ID"+
                ",(select t1.code from client t1 where t1.id = client_id) CLIENT_NAME"+
                ",(select t.name from bpartner t where t.id = c.cust_bpartner_id) CUST_BPARTNER_NAME"+
           " from BP_CONTRACT c"+
        " where "+
     "      c.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void doCustomAction(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("SUPP_BPARTNER_NAME", new FieldDef("STRING"));
	  this.fields.put("BPCONTR_NAME", new FieldDef("STRING"));
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("DOC_NO", new FieldDef("STRING"));
	  this.fields.put("DOC_DATE", new FieldDef("DATE"));
	  this.fields.put("CLIENT_ID", new FieldDef("NUMBER"));
	  this.fields.put("CUST_BPARTNER_ID", new FieldDef("NUMBER"));
	  this.fields.put("CONTRTYPE_CODE", new FieldDef("STRING"));
	  this.fields.put("STARTDATE", new FieldDef("DATE"));
	  this.fields.put("ENDDATE", new FieldDef("DATE"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("BPCONTR_ID", new FieldDef("NUMBER"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.put("CONTRSTAT_CODE", new FieldDef("STRING"));
	  this.fields.put("SUPP_BPARTNER_ID", new FieldDef("NUMBER"));
	  this.fields.put("CLIENT_NAME", new FieldDef("STRING"));
	  this.fields.put("CUST_BPARTNER_NAME", new FieldDef("STRING"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	  String[] _summaryFields = {};
	  this.summaryFields = _summaryFields;
	  this.queryResultSize = 20;
	}

}
